File: 2021/0119_getRow/0823/getRow.py
class Solution:
    def getRow(self, rowIndex: int) -> List[int]:
        res = [1] * (rowIndex + 1)
        # 从第2行开始，第0，1行均是1
        for i in range(2, rowIndex + 1):
            # 首尾元素不用计算，均为1, 其中第i行有(i + 1)个元素
            for j in range(i - 1, 0, -1):
                res[j] += res[j - 1]
        return res

File: 2021/0119_getRow/0823/getRow.rs
impl Solution {
    pub fn get_row(row_index: i32) -> Vec<i32> {
        assert!(row_index >= 0);
        let index = row_index as usize + 1;
        let mut ret = vec![1; index];
        for i in 2..index {
            for j in (1..i).rev() {
                ret[j] += ret[j - 1]; 
            }
        }
        ret

        // method 2
        assert!(row_index >= 0);
        gen_pascal_triangle()
            .skip(row_index as usize)
            .next()
            .unwrap()
    }
}

fn gen_pascal_triangle() -> impl Iterator<Item = Vec<i32>> {
    std::iter::successors(Some(vec![1]), |row| {
        vec![1].into_iter()
            .chain(row.windows(2).map(|v| v.iter().sum()))
            .chain(vec![1].into_iter())
            .collect::<Vec<i32>>()
            .into()
    })
}

